### PR TITLE
Patch flux to use flash attention 3

### PR DESCRIPTION
**Pull request by ShariqM**
Improve Flux example to use Flash Attention 3.

### Type of Change

- [ ] New example
- [x] Example updates (Bug fixes, new features, etc.)
- [ ] Other (changes to the codebase, but not to examples)

## Checklist

- [x] Example is testable in synthetic monitoring system, or `lambda-test: false` is added to example frontmatter (`---`)
  - [x] Example is tested by executing with `modal run` or an alternative `cmd` is provided in the example frontmatter (e.g. `cmd: ["modal", "deploy"]`)
  - [x] Example is tested by running with no arguments or the `args` are provided in the example frontmatter (e.g. `args: ["--prompt", "Formula for room temperature superconductor:"]`
- [x] Example is documented with comments throughout, in a [_Literate Programming_](https://en.wikipedia.org/wiki/Literate_programming) style.
- [x] Example does _not_ require third-party dependencies to be installed locally
- [x] Example pins its dependencies
  - [x] Example pins container images to a stable tag, not a dynamic tag like `latest`
  - [x] Example specifies a `python_version` for the base image, if it is used
  - [x] Example pins all dependencies to at least minor version, `~=x.y.z` or `==x.y`
  - [x] Example dependencies with `version < 1` are pinned to patch version, `==0.y.z`

## Outside contributors

You're great! Thanks for your contribution.

### CODE REVIEW

File: 06_gpu_and_ml/stable_diffusion/flux.py [hunk 1]
```diff
@@ -94,3 +196,20 @@ def main(
     print(f"Saving it to {output_path}")
     with open(output_path, "wb") as f:
         f.write(image_bytes)
+
+
+# That's it! We can now run our `main` function to generate an image. You should
```
Comment (charlesfrye) [inline 06_gpu_and_ml/stable_diffusion/flux.py:201]: I would remove this section at the bottom in favor of a more abstract description of what folks will see. And you can put that information above the function, where you say "by default we hit generate an image twice" (which I think is a typo?).

File: 06_gpu_and_ml/stable_diffusion/flux.py [hunk 2]
```diff
@@ -3,20 +3,51 @@
...
 #
-# Thanks to [@Arro](https://github.com/Arro) for the original contribution.
+# To make this example go [brrrr](https://horace.io/brrr_intro.html) we will use
+# the latest version of [Flash Attention](https://github.com/Dao-AILab/flash-attention?tab=readme-ov-file)
+# which significanly reduces the latency of the Query, Key, Value (QKV)
+# operations of Transformer models on GPUs. The key insight of Flash Attention
+# is to break the QKV operations into tiny blocks that can be computed quickly
+# in high bandwidth SRAM rather than constantly using low bandwidth VRAM. Read
+# more [here](https://gordicaleksa.medium.com/eli5-flash-attention-5c44017022ad).
+# Note that Flash Attention 3 (FA3) is specifically designed for H100 GPUs.
```
Comment (charlesfrye) [inline 06_gpu_and_ml/stable_diffusion/flux.py:19]:
```suggestion
# Note that Flash Attention 3 (FA3), which we use here, is specifically designed for H100 GPUs.
```

File: 06_gpu_and_ml/stable_diffusion/flux.py [hunk 3]
```diff
@@ -3,20 +3,51 @@
...
+# We start off by importing `BytesIO` for return image byte streams,
+# along with `Path` and `modal`.
+
+import time
 from io import BytesIO
 from pathlib import Path
 
 import modal
 
+# We'll use the schnell variant of Flux.1 which is faster but less accurate.
```
Comment (charlesfrye) [inline 06_gpu_and_ml/stable_diffusion/flux.py:30]:
```suggestion
# We'll use the `schnell` variant of Flux.1 which is the fastest but lowest quality model in the series.
```

File: 06_gpu_and_ml/stable_diffusion/flux.py [hunk 4]
```diff
@@ -3,20 +3,51 @@
...
+# If you want to use the dev variant, get yourself a Hugging Face  API key,
+# put it in your [Secrets](https://modal.com/docs/guide/secrets) and include
```
Comment (charlesfrye) [inline 06_gpu_and_ml/stable_diffusion/flux.py:32]:
```suggestion
# put it in your Modal [Secrets](https://modal.com/docs/guide/secrets) and include
```

File: 06_gpu_and_ml/stable_diffusion/flux.py [hunk 5]
```diff
@@ -3,20 +3,51 @@
...
+# along with `Path` and `modal`.
+
+import time
 from io import BytesIO
 from pathlib import Path
 
 import modal
 
+# We'll use the schnell variant of Flux.1 which is faster but less accurate.
+# If you want to use the dev variant, get yourself a Hugging Face  API key,
```
Comment (charlesfrye) [inline 06_gpu_and_ml/stable_diffusion/flux.py:31]:
```suggestion
# If you want to use the dev variant, get yourself a Hugging Face API key,
```

File: 06_gpu_and_ml/stable_diffusion/flux.py [hunk 6]
```diff
@@ -36,42 +67,95 @@
         "safetensors==0.4.4",
         "sentencepiece==0.2.0",
     )
+    # Since Hugging Face does not support Flash Attention 3 (FA3) yet (Sep 2024),
```
Comment (charlesfrye) [inline 06_gpu_and_ml/stable_diffusion/flux.py:70]:
```suggestion
    # Since Hugging Face does not support Flash Attention 3 as of writing (Sep 2024),
```

File: 06_gpu_and_ml/stable_diffusion/flux.py [hunk 7]
```diff
@@ -36,42 +67,95 @@
...
 app = modal.App("example-flux")
 
 with flux_image.imports():
     import torch
     from diffusers import FluxPipeline
 
 
+# Now we can define our `Model` class which will
+# 1) Download the model from Hugging Face Hub
+# 2) Optimize the Flux's internal Transformer and VAE models to use FA3
```
Comment (charlesfrye) [inline 06_gpu_and_ml/stable_diffusion/flux.py:99]:
```suggestion
# 2) Swap in FA3 in Flux's internal Transformer and VAE models
```

File: 06_gpu_and_ml/stable_diffusion/flux.py [hunk 8]
```diff
@@ -36,42 +67,95 @@
...
+# 3) Call `torch.compile() `with the `max-autotune` mode to optimize the models
```
Comment (charlesfrye) [inline 06_gpu_and_ml/stable_diffusion/flux.py:100]:
```suggestion
# 3) Call `torch.compile()` in `max-autotune` mode to optimize the models
```

File: 06_gpu_and_ml/stable_diffusion/flux.py [hunk 9]
```diff
@@ -80,11 +164,29 @@ def inference(self, prompt):
         return byte_stream.getvalue()
 
 
+# Finally we write our `main` entrypoint to  utilize the `Model`
```
Comment (charlesfrye) [inline 06_gpu_and_ml/stable_diffusion/flux.py:167]:
```suggestion
# Finally we write our `main` entrypoint to use the `Model`
```

**Review by charlesfrye — CHANGES_REQUESTED** [review]

File: 06_gpu_and_ml/stable_diffusion/flux.py (hunk 1 above)
Comment (ShariqM) [reply 06_gpu_and_ml/stable_diffusion/flux.py]: next version ready for you to review!

**Comment by charlesfrye** [comment]
see #904
